[SecuritySolution] Fix building block alert highlighting

janmonschke opened the pull request:
## Summary

As described in https://github.com/elastic/kibana/issues/152318, we noticed that building block alerts were not highlighted anymore after the migration to the new alerts table.

A preferred implementation of building block alert highlighting would follow the [`EUIDataGrid` approach of row highlighting](https://eui.elastic.co/#/tabular-content/data-grid-style-display#grid-row-classes). The `DataGrid` allows you to pass custom CSS class names for each row (`gridStyle.rowClasses`). That would allow us to highlight table rows with building block alerts.

However, without access to the underlying data, we would not be able generate the correct `rowClasses` for rows with building block alerts. So simply passing `gridStyle.rowClasses` to the `AlertsStateTable` was not an option.

Therefore in this PR we're introducing a new prop on the `AlertsTable`, the `highlightedRowMapper`. It's a callback function that receives the alert data and when it returns true, that row will be highlighted.

This allows for highlighting of rows from the outside without exposing too many details about the underlying data structures.

**Screenshot of the alerts table with a highlightedRowMapper that highlights building block alerts**

<img width="1259" alt="Screenshot 2023-04-21 at 13 03 54" src="https://user-images.githubusercontent.com/68591/233620704-a56204c0-e285-4289-897a-58481f440446.png">

### Additional notes

- Since the alerts table has default grid styles, it allows to pass `gridStyle` and it computes its own `rowClasses` for "active row" highlighting, the logic for merging all those styles looks intimidating. I tried my best to comment that part of code to make it clear why the merges are necessary and how they work.
- While working on the issue, I noticed that active rows are not highlighted anymore (related bug: https://github.com/elastic/kibana/issues/155487). The changes in this PR fix that behaviour as well as you can see in the screenshot below:

<img width="936" alt="Screenshot 2023-04-21 at 13 04 15" src="https://user-images.githubusercontent.com/68591/233620752-d752dada-9c97-4f00-933a-5425e19a5793.png">

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


<!--ONMERGE {"backportTargets":["8.8"]} ONMERGE-->

janmonschke committed:
feat: Add rowClassMapper to AlertsTable

janmonschke committed:
fix: bring back building block alert highlighting

janmonschke committed:
chore: fix indentation

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

lgestc commented on x-pack/plugins/security_solution/public/detections/components/alerts_table/index.tsx:
```diff
@@ -244,6 +247,13 @@ export const AlertsTableComponent: FC<DetectionEngineAlertTableProps> = ({
+  // Highlight rows with building block alerts
+  const rowClassMapper: AlertsTableStateProps['rowClassMapper'] = useCallback(
+    (alert: Alert) =>
+      alert[ALERT_BUILDING_BLOCK_TYPE] ? 'alertsTable-rowBackgroundHighlighted' : '',
```
is there a way to import this class name from `alerts_table/index.scss`?

lgestc replied:
like its done here: `x-pack/plugins/canvas/shareable_runtime/components/page.tsx`

janmonschke replied:
Ohhh, that's neat. I didn't know we could use CSS modules in Kibana.

Though, I just tried to transform it to a CSS module and it doesn't seem to work. Will ask the core team if I need to change something to get CSS modules to work properly in `security_solution` 👍

janmonschke replied:
Moved the class to the alert table package now (`plugins/triggers_actions_ui/public/application/sections/alerts_table/alerts_table.scss`).

Sadly I couldn't make CSS modules work in that context (or in any other for that matter)

lgestc commented on x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_table/alerts_table.tsx:
```diff
@@ -377,6 +379,68 @@ const AlertsTable: React.FunctionComponent<AlertsTableProps> = (props: AlertsTab
+          return rowClasses;
+        },
+        {}
+      );
+    }
+
+    // If ANY additional rowClasses have been provided, we need to merge them
+    if (propGridStyle.rowClasses || mappedRowClasses) {
+      // Get all row indices with a rowClass.
+      const mergedKeys = [
+        ...Object.keys(activeRowClasses),
+        ...Object.keys(propGridStyle.rowClasses || {}),
+        ...Object.keys(mappedRowClasses || {}),
+      ];
+      // Deduplicate keys to avoid extra iterations
+      const dedupedKeys = Array.from(new Set(mergedKeys));
+
+      // For each index, merge row classes
+      const mergedRowClasses = dedupedKeys.reduce<NonNullable<EuiDataGridStyle['rowClasses']>>(
+        (rowClasses, key) => {
+          const intKey = parseInt(key, 10);
+
+          // Concatenate class names for active row class and passed-in row class
+          // (if either exists)
+          rowClasses[intKey] = [
+            activeRowClasses[intKey],
+            propGridStyle.rowClasses?.[intKey],
+            mappedRowClasses?.[intKey],
+          ]
+            .filter(Boolean)
+            .join(' ');
+          return rowClasses;
+        },
+        {}
+      );
+      mergedGridStyle.rowClasses = mergedRowClasses;
+    }
+    return mergedGridStyle;
+  }, [alerts, activeRowClasses, props.rowClassMapper, pagination, props.gridStyle]);
+
```
it would probably be easier to test then

janmonschke replied:
Agree, will move this piece of code to its own file. 

For testing however I would like to test the public interface instead of the internals. That would allow us to make sure the integration with the Data Grid doesn't break in the future.

I am talking to ResponseObs to see why the tests are not rendering a proper Data Grid. Once that's sorted, we should be able to write proper tests.

In case the test renderer is not easy to fix, I can still come back and add unit tests for only this part of the logic. Wdyt?

janmonschke replied:
With the latest "refactor" this has split up a bit. Do you still think it makes sense to movie it outside now?

lgestc reviewed (commented):
some suggestions / question about css modules added :)

janmonschke committed:
refactor: rowClassMapper -> highlightedRowMapper

janmonschke committed:
test: assert that rows with building block alerts are highlighted

logeekal commented on x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_table/alerts_table.tsx:
```diff
@@ -286,11 +286,31 @@ const AlertsTable: React.FunctionComponent<AlertsTableProps> = (props: AlertsTab
+    if (highlightedRowMapper) {
+      mappedRowClasses = alerts.reduce<NonNullable<EuiDataGridStyle['rowClasses']>>(
+        (rowClasses, alert, index) => {
+          const shouldHighlightRow = highlightedRowMapper(alert);
```
`nit`: By name `highlightedRowMapper` seems that it is mapping the `rowclasses` ( what you were doing previously). Could we rename to something like `shouldHighlightRow`?

janmonschke replied:
Great idea 👍 https://github.com/elastic/kibana/pull/155497/commits/568b1ffdcbb11d08f2733ddf9ac05e263cbe668f

logeekal commented on x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_table/alerts_table.tsx:
```diff
@@ -377,6 +397,50 @@ const AlertsTable: React.FunctionComponent<AlertsTableProps> = (props: AlertsTab
+
+          // Concatenate class names for active row class and passed-in row class
+          // (if either exists)
+          rowClasses[intKey] = [activeRowClasses[intKey], propGridStyle.rowClasses?.[intKey]]
```
Should it also contain the `highlightedRowClasses`?  Are we saying that if `propGridStyle.rowClasses` is passed, `highlightedRowClasses` will not be taking into consideration?

logeekal replied:
Yep.. Even if I give rowClass for only one row in `propsGridStyle`, it ignore highlighted rowClasses entirely. Here is the video.

https://user-images.githubusercontent.com/7485038/234807272-2eac0752-1f60-4b41-bfd6-e065c492a85e.mov

janmonschke replied:
Nice catch, fixed in https://github.com/elastic/kibana/pull/155497/commits/fed125a6e41b763562df8b5432e300f9013869d0 and https://github.com/elastic/kibana/pull/155497/commits/77ea66dd74f842198133a067c21f4ca9a1e1de45.

It makes an assumption though that internal classes have precedence over custom ones. That made sense to me, but maybe @jcger has a different opinion?

logeekal replied:
I agree with that.

logeekal reviewed (commented):
In summary: 

- Building block alert is working great.
- It does not work if I pass `gridStyle` in rowClasses
- Active alert is not being highlighted.

https://user-images.githubusercontent.com/7485038/234808791-6f6a35ed-8fff-4f83-b957-b33d7a7d3af1.mov

janmonschke committed:
refactor: highlightedRowMapper -> shouldHighlightRow

janmonschke committed:
fix: merge row classes correctly

janmonschke committed:
fix: use internal row classes over custom ones

Also, we cannot join classes with a space string, since the data grid internally uses `classList.add` which breaks for classes with spaces.

janmonschke commented:
Fixed all the issues @logeekal 

>     * Active alert is not being highlighted.

This is only true for the alert table in security, since we are using our custom flyouts. That didn't work before either.

janmonschke committed:
chore: remove console.log statement

janmonschke commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into bug-building-block-alert-highlighting

logeekal commented on x-pack/plugins/security_solution/cypress/tasks/rule_details.ts:
```diff
@@ -158,3 +159,7 @@ export const hasIndexPatterns = (indexPatterns: string) => {
   cy.get(EDIT_RULE_SETTINGS_LINK).click();
 };
+
+export const assertRowsAreHighlighted = () => {
```
`non-issue` - Glo discourages to put assertions in the tasks and emphasizes to be directly in the tests, specially since it is one line.

Honestly, I am not sure why is that? May a convention. So I will leave it upto you.

janmonschke replied:
Haha, didn't know that. I added the assertion here since there is another assertion in this file. But will switch to the inline check. It's more concise imo. Thanks!

logeekal reviewed (approved):
LGTM 👌 . Great solution to highlight problem.

janmonschke committed:
refactor: assert inline, rather than as a task

jcger reviewed (approved):
nice!

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.8|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/156137">](https://github.com/elastic/kibana/pull/156137)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->
